src/components/app/UrlManager.js

```diff
@@ -96,17 +96,20 @@ class UrlManager extends React.PureComponent<Props> {
 
     try {
       // Process the raw url and fetch the profile.
-      const results: {
-        profile: Profile | null,
-        shouldSetupInitialUrlState: boolean,
+      // We try to fetch the profile before setting the url state, because
+      // while processing and especially upgrading the url information we may
+      // need the profile data.
+      //
+      // Also note the profile may be null for the `from-addon` dataSource since
+      // we do not `await` for retrieveProfileFromAddon function, but also in
+      // case of fatal errors in the process of retrieving and processing a
+      // profile. To handle the latter case properly, we won't `pushState` if
+      // we're in a FATAL_ERROR state.
+      const {
+        profile,
+        shouldSetupInitialUrlState,
       } = await getProfilesFromRawUrl(window.location);
 
-      // Manually coerce these into the proper type due to the FlowFixMe above.
-      // Profile may be null only for the `from-addon` dataSource since we do
-      // not `await` for retrieveProfileFromAddon function.
```

Comment:
I removed all the coercion pieces because there's no $FlowFixMe anymore. I checked that flow was still checking stuff by adding errors.